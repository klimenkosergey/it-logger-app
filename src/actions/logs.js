export const getLogs = () => async dispatch => {
  try {
    const res = await fetch('/logs');
    const data = await res.json();

    dispatch({ type: 'GET_LOGS', payload: data });
  } catch (error) {
    dispatch({ type: 'SET_ERROR', payload: 'Error loading logs' });
  }
};

export const addLog = log => async dispatch => {
  try {
    const res = await fetch('/logs', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(log)
    });
    const data = await res.json();

    dispatch({ type: 'ADD_LOG', payload: data });
  } catch (error) {
    dispatch({ type: 'SET_ERROR', payload: 'Error adding log' });
  }
};

export const editLog = log => async dispatch => {
  try {
    const res = await fetch(`/logs/${log.id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(log)
    });
    const data = await res.json();

    dispatch({ type: 'EDIT_LOG', payload: data });
  } catch (error) {
    dispatch({ type: 'SET_ERROR', payload: 'Error editing log' });
  }
};

export const deleteLog = id => async dispatch => {
  try {
    await fetch(`/logs/${id}`, {
      method: 'DELETE'
    });

    dispatch({ type: 'DELETE_LOG', payload: id });
  } catch (error) {
    dispatch({ type: 'SET_ERROR', payload: 'Error removing log' });
  }
};

export const searchLogs = text => async dispatch => {
  try {
    const res = await fetch(`/logs/?q=${text}`);
    const data = await res.json();

    dispatch({ type: 'SEARCH_LOGS', payload: data });
  } catch (error) {
    dispatch({ type: 'SET_ERROR', payload: 'Error searching logs' });
  }
};

export const setCurrent = log => ({ type: 'SET_CURRENT', payload: log });
