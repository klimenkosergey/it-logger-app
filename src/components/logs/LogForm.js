import React, { Fragment, useState, useEffect } from 'react';

import TechSelectOptions from '../techs/TechSelectOptions';

const LogForm = props => {
  const [message, setMessage] = useState('');
  const [tech, setTech] = useState('');
  const [attention, setAttention] = useState(false);

  useEffect(() => {
    setMessage(props.message ? props.message : message);
    setTech(props.tech ? props.tech : tech);
    setAttention(props.attention ? props.attention : attention);
    // eslint-disable-next-line
  }, [props]);

  const onSubmit = e => {
    e.preventDefault();
    if (props.id) {
      props.onSubmit(props.id, message, tech, attention, new Date());
    } else {
      props.onSubmit(message, tech, attention, new Date());
    }

    // Clear inputs
    setMessage('');
    setTech('');
    setAttention(false);
  };

  return (
    <Fragment>
      <div className='modal-content'>
        <h4 className='center'>
          {!props.message && !props.tech ? 'Add' : 'Edit'} System Log
        </h4>
        <div className='row'>
          <div className='input-field'>
            <input
              type='text'
              name='message'
              value={message}
              onChange={e => setMessage(e.target.value)}
            />
            {!props.id && (
              <label htmlFor='message' className='active'>
                Log Message
              </label>
            )}
          </div>
        </div>
        <div className='row'>
          <div className='input-field'>
            <select
              name='tech'
              value={tech}
              className='browser-default'
              onChange={e => setTech(e.target.value)}
            >
              <option value='' disabled>
                Select Technician
              </option>
              <TechSelectOptions />
            </select>
          </div>
        </div>
        <div className='row'>
          <div className='input-field'>
            <label>
              <input
                type='checkbox'
                className='filled-in'
                checked={attention}
                value={attention}
                onChange={e => setAttention(!attention)}
              />
              <span>Needs Attention</span>
            </label>
          </div>
        </div>
      </div>
      <div className='modal-footer'>
        <a href='#!' onClick={onSubmit} className='waves-light blue btn'>
          Submit
        </a>
      </div>
    </Fragment>
  );
};

export default LogForm;
