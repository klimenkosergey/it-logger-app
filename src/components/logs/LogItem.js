import React from 'react';
import PropTypes from 'prop-types';
import Moment from 'react-moment';

const LogItem = ({ log, deleteLog, setCurrent }) => (
  <li className='collection-item'>
    <a
      href='#edit-log-modal'
      className={`modal-trigger ${log.attention ? 'red-text' : 'blue-text'}`}
      onClick={e => setCurrent(log)}
    >
      {log.message}
    </a>
    <br />
    <span className='grey-text'>
      <span className='black-text'>ID #{log.id}</span> last edited by{' '}
      <span className='black-text'>{log.tech}</span> on{' '}
      <Moment format={'MMMM Do YYYY, hh:mm:ss a'}>{log.date}</Moment>
    </span>
    <a
      href='#!'
      onClick={e => {
        e.preventDefault();
        deleteLog(log.id);
      }}
      className='secondary-content'
    >
      <i className='material-icons grey-text'>delete</i>
    </a>
  </li>
);

LogItem.propTypes = {
  log: PropTypes.object.isRequired,
  deleteLog: PropTypes.func.isRequired
};

export default LogItem;
