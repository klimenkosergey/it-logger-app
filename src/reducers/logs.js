const logsReducerDefaultState = {
  logs: [],
  current: null,
  loading: true,
  error: null
};

const logsReducer = (state = logsReducerDefaultState, action) => {
  switch (action.type) {
    case 'GET_LOGS':
      return {
        ...state,
        loading: false,
        logs: action.payload
      };
    case 'SEARCH_LOGS':
      return {
        ...state,
        logs: action.payload
      };
    case 'ADD_LOG':
      return {
        ...state,
        logs: [...state.logs, action.payload]
      };
    case 'EDIT_LOG':
      return {
        ...state,
        logs: state.logs.map(log =>
          log.id === action.payload.id ? action.payload : log
        )
      };
    case 'DELETE_LOG':
      return {
        ...state,
        logs: state.logs.filter(log => log.id !== action.payload)
      };
    case 'SET_CURRENT':
      return {
        ...state,
        current: action.payload
      };
    case 'SET_ERROR':
      return {
        ...state,
        loading: false,
        error: action.payload
      };
    default:
      return state;
  }
};

export default logsReducer;
