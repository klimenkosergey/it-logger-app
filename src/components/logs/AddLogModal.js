import React from 'react';
import { connect } from 'react-redux';
import M from 'materialize-css/dist/js/materialize';

import { addLog } from '../../actions/logs';
import LogForm from './LogForm';

const AddLogModal = ({ addLog }) => {
  const onSubmit = (message, tech, attention, date) => {
    const modalElem = document.querySelector('#add-log-modal');
    const instance = M.Modal.getInstance(modalElem);

    if (message === '' || tech === '') {
      M.toast({ html: 'Please enter a message and tech' });
    } else {
      // Close the Modal
      instance.close();
      addLog({ message, tech, attention, date });
      M.toast({ html: 'Log added' });
    }
  };

  return (
    <div id='add-log-modal' className='modal'>
      <LogForm onSubmit={onSubmit} />
    </div>
  );
};

const ConnectedAddLogModal = connect(
  null,
  { addLog }
)(AddLogModal);

export { AddLogModal, ConnectedAddLogModal as default };
