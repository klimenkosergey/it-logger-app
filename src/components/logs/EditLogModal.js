import React from 'react';
import { connect } from 'react-redux';
import M from 'materialize-css/dist/js/materialize';

import { editLog } from '../../actions/logs';
import LogForm from './LogForm';

const EditLogModal = ({ editLog, current }) => {
  const onSubmit = (id, message, tech, attention, date) => {
    const modalElem = document.querySelector('#edit-log-modal');
    const instance = M.Modal.getInstance(modalElem);

    if (message === '' || tech === '') {
      M.toast({ html: 'Please enter a message and tech' });
    } else {
      instance.close();
      editLog({ id, message, tech, attention, date });
    }
  };

  return (
    <div id='edit-log-modal' className='modal'>
      <LogForm onSubmit={onSubmit} {...current} />
    </div>
  );
};

const mapStateToProps = state => ({
  current: state.logs.current
});

const ConnectedEditLogModal = connect(
  mapStateToProps,
  { editLog }
)(EditLogModal);

export { EditLogModal, ConnectedEditLogModal as default };
