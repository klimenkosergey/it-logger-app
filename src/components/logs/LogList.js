import React, { useEffect } from 'react';
import { connect } from 'react-redux';

import { getLogs, deleteLog, setCurrent } from '../../actions/logs';
import Preloader from '../layout/Preloader';
import LogItem from './LogItem';

const LogList = ({ logs, loading, getLogs, deleteLog, setCurrent }) => {
  useEffect(() => {
    getLogs();
    // eslint-disable-next-line
  }, []);

  if (loading) {
    return <Preloader />;
  } else {
    return (
      <ul className='collection with-header'>
        <li className='collection-header'>
          <h4 className='center'>System Logs</h4>
        </li>
        {logs.length === 0 ? (
          <p className='center'>No logs to show...</p>
        ) : (
          logs.map(log => (
            <LogItem
              key={log.id}
              log={log}
              deleteLog={deleteLog}
              setCurrent={setCurrent}
            />
          ))
        )}
      </ul>
    );
  }
};

const mapStateToProps = state => ({
  logs: state.logs.logs,
  loading: state.logs.loading
});

const ConnectedLogList = connect(
  mapStateToProps,
  { getLogs, deleteLog, setCurrent }
)(LogList);

export { LogList, ConnectedLogList as default };
