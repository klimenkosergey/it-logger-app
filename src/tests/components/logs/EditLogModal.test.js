import React from 'react';
import { shallow } from 'enzyme';

import { EditLogModal } from '../../../components/logs/EditLogModal';

let wrapper, editLog, current;

beforeEach(() => {
  editLog = jest.fn();
  current = {
    id: 1,
    message: 'Fixed hard drive',
    tech: 'Tony Stark',
    date: '2019-08-15T10:27:16.427Z'
  };
  wrapper = shallow(<EditLogModal editLog={editLog} current={current} />);
});

it('renders EditLogModal correctly', () => {
  expect(wrapper).toMatchSnapshot();
});
