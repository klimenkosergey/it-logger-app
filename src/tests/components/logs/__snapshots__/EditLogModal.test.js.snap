// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders EditLogModal correctly 1`] = `
<div
  className="modal"
  id="edit-log-modal"
>
  <LogForm
    date="2019-08-15T10:27:16.427Z"
    id={1}
    message="Fixed hard drive"
    onSubmit={[Function]}
    tech="Tony Stark"
  />
</div>
`;
